edb/edgeql/compiler/context.py

```diff
@@ -285,6 +285,14 @@ def __init__(
         path_scope: Optional[irast.ScopeTreeNode] = None,
         alias_result_view_name: Optional[s_name.QualName] = None,
         options: Optional[GlobalCompilerOptions] = None,
+        set_types: Optional[Dict[irast.Set, s_types.Type]] = None,
+        scope_tree_nodes: Optional[
+            MutableMapping[int, irast.ScopeTreeNode]
+        ] = None,
+        type_rewrites: Optional[dict[
+            Tuple[s_types.Type, bool],
+            irast.Set | None | Literal[True]
+        ]] = None,
```

Comment:
This is missing a lot of fields that need to be preserved in Environment for correct operation in some case or another, especially once DML is supported.
Definitely needed:
- materialized_sets
- dml_exprs
- dml_stmts
- dml_rewrites
- schema_refs

I think possibly needed, or needed if things are done a little differently:
- pointer_derivation_map (for card inference?)
- pointer_specified_info (for card inference?)
- query_globals

Plus, I think having all of the caches shared will help get more cache hits, even if it isn't *necessary*.
And I think many of the other things wouldn't hurt to have be shared.

So I think instead of constructing a fully new env, we should probably make a straight copy of the env and just update the few fields we care about changing.

It might /just/ be `options` we need to change, though? In which case it might actually just be best to update the `options` field and then restore it after.